should check GPU capabilities and exit gracefully

I think Godot should somehow check if the system can handle GLES2 or not, and instead of aborting or failing, present a meaningful error message to the user about their drivers not supporting GLES2. "Your computer does not appear to currently support GLES2, which is required to run this program. New drivers or hardware upgrade is required." Something along those lines.

I am not sure what this would take to do, but there must be some way to do it in the code.

This is specifically for Windows users who seem to always have crappy GPUs and crappy drivers, but ideally the checks would be on all platforms.

```
In file included from drivers\angle/common/angleutils.h:12:0,
                 from drivers\angle\common\angleutils.cpp:7:
drivers\angle/common/platform.h:62:28: fatal error: d3d11_1.h: No such file or directory
 #       include <d3d11_1.h>
                            ^
```
Fullstop?

Well that's the point of angle, it uses direct 3d instead of opengl :p

On 15 February 2016 at 18:57, anatoly techtonik <notifications@github.com>
wrote:

> In file included from drivers\angle/common/angleutils.h:12:0,
>                  from drivers\angle\common\angleutils.cpp:7:
> drivers\angle/common/platform.h:62:28: fatal error: d3d11_1.h: No such file or directory
>  #       include <d3d11_1.h>
>                             ^
>
> Fullstop?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/godotengine/godot/issues/1162#issuecomment-184411949>.
>

Lots of errors with newer gcc 5.3.0, for example:
```
In file included from drivers\angle\common\angleutils.cpp:7:0:
drivers\angle/common/angleutils.h:29:21: warning: defaulted and deleted functions only available with -std=c++11 or -std=gnu++11
     NonCopyable() = default;
                     ^
...
```

```
drivers\angle\libANGLE\renderer\d3d\d3d11\win32\NativeWindow.cpp:15:19: fatal error:
dcomp.h: No such file or directory
compilation terminated.
```
[Need](https://sourceforge.net/p/mingw-w64/mailman/mingw-w64-public/thread/56B8A51F.2070509%40codeweavers.com/#msg34832456) to wait for [new MinGW release](https://sourceforge.net/p/mingw-w64/mailman/mingw-w64-public/thread/56B8A51F.2070509%40codeweavers.com/#msg34832456).

--- a/drivers/gles2/rasterizer_gles2.cpp
+++ b/drivers/gles2/rasterizer_gles2.cpp
@@ -10788,8 +10788,17 @@ void RasterizerGLES2::init() {
 	if (OS::get_singleton()->is_stdout_verbose()) {
 		print_line(String("GLES2: Using GLEW ") + (const char*) glewGetString(GLEW_VERSION));
 	}
-#endif
 
+	// Check for GL 2.1 compatibility, if not bail out
+	if (!glewIsSupported("GL_VERSION_2_1")) {
+		ERR_PRINT("Your system's graphic drivers seem not to support OpenGL 2.1 / GLES 2.0, sorry :(\n"
+			  "Try a drivers update, buy a new GPU or try software rendering on Linux; Godot will now crash with a segmentation fault.");
+		OS::get_singleton()->alert("Your system's graphic drivers seem not to support OpenGL 2.1 / GLES 2.0, sorry :(\n"
+					   "Godot Engine will self-destruct as soon as you acknowledge this error message.",
+					   "Fatal error: Insufficient OpenGL / GLES drivers");
+		// TODO: If it's even possible, we should stop the execution without segfault and memory leaks :)
+	}
+#endif
 
 
 

--- a/platform/windows/os_windows.cpp
+++ b/platform/windows/os_windows.cpp
@@ -1343,7 +1343,7 @@ void OS_Windows::vprint(const char* p_format, va_list p_list, bool p_stderr) {
 void OS_Windows::alert(const String& p_alert,const String& p_title) {
 
 	if (!is_no_window_mode_enabled())
-		MessageBoxW(NULL,p_alert.c_str(),p_title.c_str(),MB_OK|MB_ICONEXCLAMATION);
+		MessageBoxW(NULL, p_alert.c_str(), p_title.c_str(), MB_OK | MB_ICONEXCLAMATION | MB_TASKMODAL);
 	else
 		print_line("ALERT: "+p_alert);
 }

